### ISSUE TITLE
adding back a failed node crashed etcd (v2.0.10)

### ISSUE THREAD

**Issue by difro** [coreos/etcd#2746]
Adding back a failed node (with a fresh, empty data-dir) to a 6-node cluster failed 
with following log trace.

```
2015/04/23 08:45:37 etcd: listening for peers on http://0.0.0.0:2380
2015/04/23 08:45:37 etcd: listening for client requests on http://0.0.0.0:2379
2015/04/23 08:45:37 etcdserver: datadir is valid for the 2.0.1 format
2015/04/23 08:45:37 etcdserver: name = cne069
2015/04/23 08:45:37 etcdserver: data dir = /etcd-data
2015/04/23 08:45:37 etcdserver: member dir = /etcd-data/member
2015/04/23 08:45:37 etcdserver: heartbeat = 100ms
2015/04/23 08:45:37 etcdserver: election = 1000ms
2015/04/23 08:45:37 etcdserver: snapshot count = 10000
2015/04/23 08:45:37 etcdserver: advertise client URLs = http://cne069:2379
2015/04/23 08:45:37 etcdserver: start member b2d62e962fccdc77 in cluster ea2e39675334c1a3
2015/04/23 08:45:37 raft: b2d62e962fccdc77 became follower at term 0
2015/04/23 08:45:37 raft: newRaft b2d62e962fccdc77 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2015/04/23 08:45:37 raft: b2d62e962fccdc77 became follower at term 1
2015/04/23 08:45:37 raft: b2d62e962fccdc77 [term: 1] received a MsgHeartbeat message with higher term from 21b659ea1139a684 [term: 5]
2015/04/23 08:45:37 raft: b2d62e962fccdc77 became follower at term 5
2015/04/23 08:45:37 raft.node: b2d62e962fccdc77 elected leader 21b659ea1139a684 at term 5
2015/04/23 08:45:37 raft: b2d62e962fccdc77 [logterm: 0, index: 21305] rejected msgApp [logterm: 5, index: 21305] from 21b659ea1139a684
2015/04/23 08:45:37 raft: b2d62e962fccdc77 [commit: 0, lastindex: 0, lastterm: 0] starts to restore snapshot [index: 20002, term: 4]
2015/04/23 08:45:37 raftlog: log [committed=1, applied=0, unstable.offset=0, len(unstable.Entries)=0] starts to restore snapshot [index: 20002, term: 4]
2015/04/23 08:45:37 raft: b2d62e962fccdc77 restored progress of 21b659ea1139a684 [next = 20003, match = 0, wait = 0]
2015/04/23 08:45:37 raft: b2d62e962fccdc77 restored progress of 4ff26076565bf354 [next = 20003, match = 0, wait = 0]
2015/04/23 08:45:37 raft: b2d62e962fccdc77 restored progress of 5d487845232dae55 [next = 20003, match = 0, wait = 0]
2015/04/23 08:45:37 raft: b2d62e962fccdc77 restored progress of 672796e3197d3607 [next = 20003, match = 0, wait = 0]
2015/04/23 08:45:37 raft: b2d62e962fccdc77 restored progress of 8f181250536b3e6a [next = 20003, match = 0, wait = 0]
2015/04/23 08:45:37 raft: b2d62e962fccdc77 restored progress of bd1cfbd83540643e [next = 20003, match = 0, wait = 0]
2015/04/23 08:45:37 raft: b2d62e962fccdc77 restored progress of f234803abb54b919 [next = 20003, match = 0, wait = 0]
2015/04/23 08:45:37 raft: b2d62e962fccdc77 [commit: 0] restored snapshot [index: 20002, term: 4]
2015/04/23 08:45:37 etcdserver: saved incoming snapshot at index 20002
2015/04/23 08:45:37 rafthttp: starting client stream to 21b659ea1139a684 at term 5
2015/04/23 08:45:37 etcdserver: recovered from incoming snapshot at index 20002
2015/04/23 08:45:38 raft: b2d62e962fccdc77 [commit: 20002] ignored snapshot [index: 20002, term: 4]
2015/04/23 08:45:38 rafthttp: unexpected removal of unknown peer '7433075612127671815'
2015/04/23 08:45:38 rafthttp: encountered error reading the client log stream: read tcp 10.20.81.228:2380: use of closed network connection
2015/04/23 08:45:38 rafthttp: client streaming to 21b659ea1139a684 at term 5 has been stopped
panic: rafthttp: unexpected removal of unknown peer '7433075612127671815'

goroutine 39 [running]:
log.Panicf(0x8d67f0, 0x31, 0xc2081194f8, 0x1, 0x1)
        /home/core/go/src/log/log.go:314 +0xd0
github.com/coreos/etcd/rafthttp.(*transport).removePeer(0xc208010380, 0x672796e3197d3607)
        /home/core/share/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/rafthttp/transport.go:152 +0x1c3
github.com/coreos/etcd/rafthttp.(*transport).RemovePeer(0xc208010380, 0x672796e3197d3607)
        /home/core/share/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/rafthttp/transport.go:136 +0x77
github.com/coreos/etcd/etcdserver.(*EtcdServer).applyConfChange(0xc208048200, 0x844ce4d4ee0002c4, 0x1, 0x672796e3197d3607, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /home/core/share/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdserver/server.go:795 +0x81e
github.com/coreos/etcd/etcdserver.(*EtcdServer).apply(0xc208048200, 0xc2081fe000, 0x51e, 0x555, 0xc2080ba1b0, 0x0, 0x0)
        /home/core/share/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdserver/server.go:699 +0x399
github.com/coreos/etcd/etcdserver.(*EtcdServer).run(0xc208048200)
        /home/core/share/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdserver/server.go:430 +0xf4c
created by github.com/coreos/etcd/etcdserver.(*EtcdServer).start
        /home/core/share/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdserver/server.go:298 +0x294

goroutine 1 [runnable]:
github.com/coreos/etcd/etcdmain.Main()
        /home/core/share/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdmain/etcd.go:102 +0x9f7
main.main()
        /home/core/share/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/main.go:29 +0x1b

goroutine 5 [syscall]:
os/signal.loop()
        /home/core/go/src/os/signal/signal_unix.go:21 +0x1f
created by os/signal.init·1
        /home/core/go/src/os/signal/signal_unix.go:27 +0x35

goroutine 11 [sleep]:
net.func·019()
        /home/core/go/src/net/dnsclient_unix.go:240 +0x5a
created by net.loadConfig
        /home/core/go/src/net/dnsclient_unix.go:269 +0x20c

goroutine 40 [runnable]:
github.com/coreos/etcd/etcdserver.(*EtcdServer).Do(0xc208048200, 0x7f0da4e27530, 0xc208070360, 0x774ce57428ec0001, 0x83e480, 0x3, 0xc2080ba720, 0x26, 0xc2080380f0, 0x43, ...)
        /home/core/share/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdserver/server.go:496 +0x5df
github.com/coreos/etcd/etcdserver.(*EtcdServer).publish(0xc208048200, 0x12a05f200)
        /home/core/share/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdserver/server.go:658 +0x30b
created by github.com/coreos/etcd/etcdserver.(*EtcdServer).Start
        /home/core/share/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdserver/server.go:279 +0x4f

goroutine 41 [runnable]:
github.com/coreos/etcd/etcdserver.(*EtcdServer).purgeFile(0xc208048200)
        /home/core/share/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdserver/server.go:309 +0x48c
created by github.com/coreos/etcd/etcdserver.(*EtcdServer).Start
        /home/core/share/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdserver/server.go:280 +0x69

goroutine 42 [IO wait]:
net.(*pollDesc).Wait(0xc20808aa00, 0x72, 0x0, 0x0)
        /home/core/go/src/net/fd_poll_runtime.go:84 +0x47
net.(*pollDesc).WaitRead(0xc20808aa00, 0x0, 0x0)
        /home/core/go/src/net/fd_poll_runtime.go:89 +0x43
net.(*netFD).accept(0xc20808a9a0, 0x0, 0x7f0da4e24c18, 0xc2080f8da0)
        /home/core/go/src/net/fd_unix.go:419 +0x40b
net.(*TCPListener).AcceptTCP(0xc208036040, 0x7f0da4e27670, 0x0, 0x0)
        /home/core/go/src/net/tcpsock_posix.go:234 +0x4e
net.(*TCPListener).Accept(0xc208036040, 0x0, 0x0, 0x0, 0x0)
        /home/core/go/src/net/tcpsock_posix.go:244 +0x4c
github.com/coreos/etcd/pkg/transport.(*rwTimeoutListener).Accept(0xc20801fa40, 0x0, 0x0, 0x0, 0x0)
        /home/core/share/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/pkg/transport/timeout_listener.go:44 +0x7a
net/http.(*Server).Serve(0xc2080704e0, 0x7f0da4e26448, 0xc20801fa40, 0x0, 0x0)
        /home/core/go/src/net/http/server.go:1728 +0x92
github.com/coreos/etcd/etcdmain.serveHTTP(0x7f0da4e26448, 0xc20801fa40, 0x7f0da4e27420, 0xc2080ba150, 0x45d964b800, 0x0, 0x0)
        /home/core/share/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdmain/http.go:36 +0x1be
github.com/coreos/etcd/etcdmain.func·005(0x7f0da4e26448, 0xc20801fa40)
        /home/core/share/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdmain/etcd.go:198 +0x5a
created by github.com/coreos/etcd/etcdmain.startEtcd
        /home/core/share/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdmain/etcd.go:199 +0x18a6

goroutine 43 [IO wait]:
net.(*pollDesc).Wait(0xc20808aae0, 0x72, 0x0, 0x0)
        /home/core/go/src/net/fd_poll_runtime.go:84 +0x47
net.(*pollDesc).WaitRead(0xc20808aae0, 0x0, 0x0)
        /home/core/go/src/net/fd_poll_runtime.go:89 +0x43
net.(*netFD).accept(0xc20808aa80, 0x0, 0x7f0da4e24c18, 0xc2080bc3a0)
        /home/core/go/src/net/fd_unix.go:419 +0x40b
net.(*TCPListener).AcceptTCP(0xc208036048, 0x0, 0x0, 0x0)
        /home/core/go/src/net/tcpsock_posix.go:234 +0x4e
net.(*TCPListener).Accept(0xc208036048, 0x0, 0x0, 0x0, 0x0)
        /home/core/go/src/net/tcpsock_posix.go:244 +0x4c
github.com/coreos/etcd/pkg/transport.(*keepaliveListener).Accept(0xc208033690, 0x0, 0x0, 0x0, 0x0)
        /home/core/share/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/pkg/transport/keepalive_listener.go:48 +0x71
net/http.(*Server).Serve(0xc208070540, 0x7f0da4e26480, 0xc208033690, 0x0, 0x0)
        /home/core/go/src/net/http/server.go:1728 +0x92
github.com/coreos/etcd/etcdmain.serveHTTP(0x7f0da4e26480, 0xc208033690, 0x7f0da4e27570, 0xc2080b2040, 0x0, 0x0, 0x0)
        /home/core/share/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdmain/http.go:36 +0x1be
github.com/coreos/etcd/etcdmain.func·006(0x7f0da4e26480, 0xc208033690)
        /home/core/share/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdmain/etcd.go:206 +0x80
created by github.com/coreos/etcd/etcdmain.startEtcd
        /home/core/share/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdmain/etcd.go:207 +0x19f2

goroutine 44 [chan receive]:
github.com/coreos/etcd/pkg/osutil.func·001()
        /home/core/share/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/pkg/osutil/interrupt_unix.go:53 +0x68
created by github.com/coreos/etcd/pkg/osutil.HandleInterrupts
        /home/core/share/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/pkg/osutil/interrupt_unix.go:74 +0x1f6

goroutine 45 [runnable]:
github.com/coreos/etcd/pkg/fileutil.func·001()
        /home/core/share/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/pkg/fileutil/purge.go:69 +0x658
created by github.com/coreos/etcd/pkg/fileutil.PurgeFile
        /home/core/share/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/pkg/fileutil/purge.go:75 +0x231

goroutine 46 [runnable]:
github.com/coreos/etcd/pkg/fileutil.func·001()
        /home/core/share/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/pkg/fileutil/purge.go:69 +0x658
created by github.com/coreos/etcd/pkg/fileutil.PurgeFile
        /home/core/share/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/pkg/fileutil/purge.go:75 +0x231
```

It was added with following args
```
 -name cne069 -data-dir '/etcd-data' 
 -initial-advertise-peer-urls http://cne069:2380
 -advertise-client-urls http://cne069:2379
 -listen-peer-urls http://0.0.0.0:2380
 -listen-client-urls http://0.0.0.0:2379
 -initial-cluster-token setcd
 -initial-cluster cne068=http://tne068:2380,tne069=http://tne069:2380,cne070=http://cne070:2380,cne064=http://cne064:2380,tne070=http://tne070:2380,cne069=http://cne069:2380,tne064=http://tne064:2380
 -initial-cluster-state existing
```

This does not happen 100% of time.
but when it does trigger a crash, even adding another new never-been-in-the-cluster node crashes.

Also, sometimes this heals itself ??
For example, what crashed yesterday will work fine after a day. (with same set of commands adding the same node to the same cluster)

**Comment by yichengq** [comment]
@difro Yes, what you observe may happen, and we are fixing it at here: https://github.com/coreos/etcd/pull/2701

**Comment by xiang90** [comment]
@difro @yichengq I can confirm this is a bug as described in #2701. It happens when the peer is in raft membership but not in the cluster membership of the newly joined member. We are solving this by making the cluster membership to be consistent with the raft membership.

**Comment by yichengq** [comment]
This has been fixed by #2753

**Comment by yichengq** [comment]
Thanks for reporting!
